trace-dispatcher/doc/trace-dispatcher.md

```diff
@@ -0,0 +1,798 @@
+
+To further prevent occasional leaks of `Confidential` traces, all output from those traces is tagged with the `CONFIDENTIAL` keyword.
+
+### Frequency
+
+__Frequency filtering__ is yet another part of __trace filtering__, and represents an optional limit on the observable frequency of individual trace messages.
+
+Semantically, this is corresponds to a randomly-fair suppression of messages within a particular trace, when their moving-average frequency exceeds a given threshold parameter.
+
+The __frequency limiter__ itself emits a __suppression summary__ message under the following conditions:
+
+* when it message suppression begins, and
+* when message suppression stops -- adding the number of suppressed messages.
+
+__Frequency limiters__ are given a name to identify its activity.
+
+```haskell
+limitFrequency
+  :: MonadIO m
+  => Double                  -- ^ messages per second
+  -> Text                    -- ^ name of this limiter
+  -> Trace m a               -- ^ the limited trace
+  -> Trace m LimitingMessage -- ^ a trace emitting the messages of the limiter
+  -> m (Trace m a)           -- ^ the original trace
+
+data LimitingMessage =
+    StartLimiting Text
+  | StopLimiting  Text Int
+```
+
+The frequency filtering is intended to be applied to a subset of traces (those known to be noisy).  For this subset of traces the frequency limit can be configured.
+
+## Presentation
+### Formatting
+
+The `LogFormatting` typeclass is used to describe __trace presentation__ -- mapping __traces__ to __metrics__ and __messages__.
+
+* The `forMachine` method is used for a machine readable representation, which can varied through detail level.
+  It requires an implementation to be provided by the trace author.
+
+* the `forHuman` method shall represent the message in human readable form.
+  It's default implementation defers to `forMachine`.
+
+* the `asMetrics` method shall represent the message as `0` to `n` metrics.
+  It's default implementation assumes no metrics. If a text is given it is
+  appended as last element to the namespace.
+
+```haskell
+class LogFormatting a where
+  forMachine :: DetailLevel -> a -> A.Object
+
+  forHuman :: a -> Text
+  forHuman = forMachine DNormal
+
+  asMetrics :: a -> [Metric]
+  asMetrics v = []
+
+data Metric
+    = IntM (Maybe Text) Int
+    | DoubleM (Maybe Text) Double
+    deriving (Show, Eq)
+```
+
+The standard formatters transform a stream of messages of `a`, where `a` is an instance of `LogFormatter` to a stream of `FormattedMessages`.
+
+```haskell
+data FormattedMessage
+    = Human Text
+    | Machine Text
+    | Metrics [Metric]
+```
+
+`humanFormatter` takes a `Bool` argument, which tells if color codes for the standard output __trace-out__ shall be inserted, and an argument which is the app name, which gets prepended to the namespace, while the `machineFormatter` has as arguments the desired detail level and as well the application name.  `metricsFormatter` takes no extra arguments:
```

Comment:
Style: you could probably move most of this paragraph into haddocks for the arguments in the code snippet bellow.

Reply:
Okay

Reply:
@jutaro I meant something like this:

```haskell
-- | Format this trace for human readability
humanFormatter :: (LogFormatting a, MonadIO m)
  => Bool
  -- ^ True if this representation is for the console and should be colored.
  -> Text
  -- ^ The application name which will be prepended to the namespace.
  -> Trace m FormattedMessage
  -> m (Trace m a)
```

That way it's more clear which argument you're talking about. But I'm not going to block you on this :-)